### DISCUSSION TITLE
[Question] How are scrambles generated?

### DISCUSSION THREAD

**Discussion by fer-hnndz** [thewca/tnoodle-lib#34] [Q&A]
Hey there, sorry for opening an issue for this, I tried to find any other resources of how are the official scrambles generated for competitions.

I'm currently working on a timer app, written in Python. The way I implemented the scramble generation is not the best nor the official way. I would like to implement official scrambles into my app and possibly make a Python module out of it. If you all are ok with it.

I don't need actual code, just an explanation of how it is done.

Thanks

**Comment by gregorbg** [comment]
Hey, thanks for reaching out! I will move this to the Discussion feature, because it's not technically an issue.

**Comment by gregorbg** [comment]
So basically, all the TNoodle scrambling code (the "mathematical" portion) lives in this `tnoodle-lib` repository. The `tnoodle` repository contains the additional code necessary for running the frontend and drawing PDFs.

Before diving into the specifics, it is important to understand that TNoodle is written in Java. If you're writing a Python app, you will have to establish Java compatibility one way or another. Check out https://github.com/thewca/tnoodle-lib/discussions/24 for some ideas and suggestions on that front.

Once you can call Java code, you probably want to take a look at the [`PuzzleRegistry`](https://github.com/thewca/tnoodle-lib/blob/master/scrambles/src/main/java/org/worldcubeassociation/tnoodle/scrambles/PuzzleRegistry.java) class. It contains cached scramblers for all puzzles which TNoodle can generate scrambles for.
Using the scramblers is really easy, and you can see a working proof-of-concept sample at https://github.com/thewca/tnoodle-samples/blob/master/java-mvp/src/main/java/com/foobar/mvp/ScrambleProvider.java

**Reply by fer-hnndz** [reply]
So as I see the scramble program uses a lot of computational resources, so definetely, I'll have to discard the idea of implementing the same logic, but in Python. Probably considering learning Java for this. The idea of running a local API that returns the scramble is not bad. But I'm afraid that it will take time to implement it. Any suggestions of what path should i take?
